### DISCUSSION TITLE
How do you resolve an ESM import from TypeScript (*.tsx) file

### DISCUSSION THREAD

**Discussion by pbadenski** [webpack/webpack#15938] [Q&A]
Typescript 4.7 requires this syntax to import ESM module from a Typescript file:
`const { installConsoleApi } = await import("@pricingmonkey/ui/src/devTools/consoleApi.mjs");`

The file itself however has a TypeScript specific extension: `consoleApi.mts`.

I know how to configure Webpack to resolve imports by trying different suffixes (via resolve.extensions), but I don't know how to replace an 'mjs' to 'mts' during resolution.

Am I doing something wrong here - I'm rather lost...

**Comment by alexander-akait** [comment]
How about it https://github.com/TypeStrong/ts-loader/issues/1383#issuecomment-968075478?

**Reply by pbadenski** [reply]
You're my hero - thank you!

Works with small adjustments:

```
class NodeNextTSPlugin {
  apply(resolver) {
    const target = resolver.ensureHook("file");

    resolver
      .getHook("raw-file")
      .tapAsync("NodeNextTSPlugin", (request, resolveContext, callback) => {
        const obj = {
          ...request,
          path: request.path.replace(/\.mjs/, '.mts'),
          relativePath:
            request.relativePath && request.relativePath.replace(/\.mjs/, '.mts')
        };

        resolver.doResolve(
          target,
          obj,
          this.appending,
          resolveContext,
          callback
        );
      });
  }
}
```

I think this should definitely be added to webpack documentation on TypeScript integration.

**Reply by alexander-akait** [reply]
@pbadenski We have request to add it in webpack core, but it is not solved :disappointed: Anyway feel free to send a PR to docs
